ablySpec/RealtimeClientChannel.swift

```diff
@@ -550,6 +550,40 @@ class RealtimeClientChannel: QuickSpec {
 
             }
 
+            // RTL7
+            context("subscribe") {
+
+                // RTL7c
+                it("should implicitly attach the channel") {
+                    let client = ARTRealtime(options: AblyTests.commonAppSetup())
+                    defer { client.close() }
+
+                    let channel = client.channels.get("test")
+
+                    channel.subscribe { _, _ in }
+
+                    expect(channel.state).toEventually(equal(ARTRealtimeChannelState.Attached), timeout: testTimeout)
+                }
+
+                // RTL7c
+                pending("should result in an error if channel is in the FAILED state") {
+                    let client = ARTRealtime(options: AblyTests.commonAppSetup())
+                    defer { client.close() }
+
+                    let channel = client.channels.get("test")
+                    channel.onError(AblyTests.newErrorProtocolMessage())
+                    expect(channel.state).to(equal(ARTRealtimeChannelState.Failed))
+
+                    waitUntil(timeout: testTimeout) { done in
+                        channel.subscribe { message, errorInfo in
```

Comment:
After #204, the "However, if the channel is in the FAILED state, the subscribe request result in an error" part is not done. The error story is not so consistent yet in this library, sometimes we throw exceptions for things like this, sometimes we move NSErrors around...

I think (and it's not an original thought) that exceptions should be used for programmer errors (e. g. bad arguments passed to a function, state violations, etc.), because those just shouldn't happen ever and if they do the code should be fixed, while "external world" errors (like a channel being FAILED) are best handled with normal "error values", like those `ErrorInfo`s, or returning nil.

In this case, I think returning a nil listener would be best; it's easy to check if you are indeed interested in whether the subscription succeeded, and it doesn't have any effect otherwise. An exception would be truly unfortunate because you would have to either always handle it or always check the channel state before subscribing, because the channel could always be in a FAILED state.